### PR TITLE
Add fungible token smart contract with few tests.

### PR DESCRIPTION
**Pull request by vivekvpandya**

### CODE REVIEW

File: examples/fungible_token/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    }
+    // TODO: Use checked math and return error if Underflow.
+    fn decrease_total_supply(&mut self, amount: u128) {
+        self.total_supply -= amount;
+    }
+    fn set_balance(&mut self, account: &ProgramId, amount: u128) {
+        self.balances.insert(*account, amount);
+    }
+    fn get_balance(&self, account: &ProgramId) -> u128 {
+        *self.balances.get(account).unwrap_or(&0u128)
```
Comment (shamilsan) [inline examples/fungible_token/src/lib.rs:61]: The suffix is not needed.
```suggestion
        *self.balances.get(account).unwrap_or(&0)
```

File: examples/fungible_token/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+            // TODO: safe math for overflow detection.
+            self.set_balance(account, old_balance + amount);
+        }
+        let transfer_data = TransferData {
+            from: H256::zero(),
+            to: H256::from_slice(account.as_slice()),
+            amount,
+        };
+        msg::reply(
+            &Event::Transfer(transfer_data),
```
Comment (shamilsan) [inline examples/fungible_token/src/lib.rs:80]: No need to get a reference here.
```suggestion
            Event::Transfer(transfer_data),
```

File: examples/fungible_token/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+        msg::reply(
+            &Event::Approval(approve_data),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+    fn get_allowance(&self, owner: &ProgramId, spender: &ProgramId) -> u128 {
+        self.allownces
+            .get(owner)
+            .map_or(0u128, |m| *m.get(spender).unwrap_or(&0u128))
```
Comment (shamilsan) [inline examples/fungible_token/src/lib.rs:160]: Make it more linear.
```suggestion
        *self.allownces
            .get(owner)
            .and_then(|m| m.get(spender))
            .unwrap_or(&0)
```

File: examples/fungible_token/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+    }
+    fn increase_allowance(&mut self, owner: &ProgramId, spender: &ProgramId, amount: u128) {
+        let allownce = self.get_allowance(owner, spender);
```
Comment (shamilsan) [inline examples/fungible_token/src/lib.rs:163]: Typo.
```suggestion
        let allowance = self.get_allowance(owner, spender);
```

File: examples/fungible_token/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+enum Event {
+    Transfer(TransferData),
+    Approval(ApproveData),
+}
+
+gstd::metadata! {
+    title: "FungibleToken",
+        init:
+            input : InitConfig,
+            output : (),
```
Comment (shamilsan) [inline examples/fungible_token/src/lib.rs:250]: No need.
```suggestion
```

File: examples/fungible_token/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        handle:
+            input : Action,
+            output : Event,
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn handle() {
+    let action: Action = msg::load().unwrap();
```
Comment (shamilsan) [inline examples/fungible_token/src/lib.rs:258]: unwrap -> expect

File: examples/fungible_token/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    fn total_supply(&self) -> u128 {
+        self.total_supply
+    }
+    #[allow(dead_code)]
+    fn decimals(&self) -> u8 {
+        18
+    }
+    // TODO: Use checked math and return error if Overflow.
+    fn increase_total_supply(&mut self, amount: u128) {
+        self.total_supply += amount;
```
[the remark this answers is not in the document]
Comment (vivekvpandya) [reply examples/fungible_token/src/lib.rs:51]: Yes it is one of TODOs:

[the remark this answers is not in the document]
Comment (shamilsan) [reply examples/fungible_token/src/lib.rs:51]: > Yes it is one of TODOs:

TODO is related to error handling. But `saturating_add` is better than nothing for now.

File: examples/Cargo.toml [hunk 8]
```diff
@@ -29,6 +29,7 @@ members = [
     "wait",
     "token-vendor",
     "token-vendor/workshop-example",
+    "fungible_token"
```
Comment (NikVolf) [inline examples/Cargo.toml:32]:
```suggestion
    "fungible-token"
```

and fixes pls

**Review by shamilsan — APPROVED** [review]

**Pull request merged**